fix: prevent eager-loaded entities with DeleteDateColumn from overwriting manual relations

LeviHeber opened the pull request:
The issue was caused by the way TypeORM processing queries. For each join, it checks if the join for that table exists. However, if the join has a condition, it will always add the join, even if a previous join for the same table already exists.

In the case of eager loading, TypeORM creates its own join, and since it has a DeleteDateColumn, an additional condition is added to the join (checking for a null DeletedDate). This newly added join does not match the manually specified join, and if the manual relation had additional nested fields that were not part of the eager load, those fields would be overwritten.

This fix addresses the issue by merging the manually requested relations with the eager-loaded data, preserving the complete details of the manual relations.

In issue #10426 , the same case was described, where there are 2 joins to the same table,
which causes performance issues.

The current fix does not change the join process and leaves both joins as they were,
but only fixes the consequences that were created as a result of it in the object creation,
because this is a consequence that is relevant both to this case and to other cases.

Unrelated, it should check how to deal with the problem of having 2 joins, for performance reasons.


Closes: #11265
Closes: #11056
Closes: #10812
Closes: #9774
Closes: #9581

<!--
  😀 Wonderful!  Thank you for opening a pull request for TypeORM.

  Please fill in the information below to expedite the review
  and (hopefully) merge of your change.

  If unsure about something.. just do as best as you're able,
  or reach out through our community support channels!
  https://github.com/typeorm/typeorm/blob/master/docs/support.md
-->

### Description of change

<!--
  Please be clear and concise what the change is intended to do,
  why this change is needed, and how you've verified that it
  corrects what you intended.

  In some cases it may be helpful to include the current behavior
  and the new behavior.

  If the change is related to an open issue, you can link it here.
  If you include `Fixes #0000` (replacing `0000` with the issue number)
  when this is merged it will automatically mark the issue as fixed and
  close it.
-->


### Pull-Request Checklist

<!--
  Please make sure to review and check all of the following.

  If an item is not applicable, you can add "N/A" to the end.
-->

- [X] Code is up-to-date with the `master` branch
- [X] `npm run format` to apply prettier formatting
- [X] `npm run test` passes with this change
- [X] This pull request links relevant issues as `Fixes #0000`
- [X] There are new or updated unit tests validating the change
- [ ] N/A - Documentation has been updated to reflect this change
- [X] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

<!--
  🎉 Thank you for contributing and making TypeORM even better!
-->

gioboa reviewed (commented):
Hi @LeviHeber, thanks for this great PR, can you fix up the sap tests please?

LeviHeber commented on test/github-issues/11265/issue-11265.ts:
```diff
@@ -0,0 +1,65 @@
+                    dataSource.getRepository(RelationEntity)
+                const parentRepository = dataSource.getRepository(ParentEntity)
+
+                const relationNested = new RelationNestedEntity()
```
I added a non auto-generated dummy field.

LeviHeber commented:
i left the test to be `describe.only`. is it need to be fixed?

gioboa reviewed (commented):
Yes, remove the .only please

LeviHeber commented:
I removed the `only` and pushed it again.
It seems that more tests are running now, but for some reason many of the tests encounter the same check (`commands - migration create: should write Javascript empty migration file when option is passed`) and it doesn't seem to be related to the fix I made.
I would appreciate some guidance.

sgarner commented:
> I removed the `only` and pushed it again. It seems that more tests are running now, but for some reason many of the tests encounter the same check (`commands - migration create: should write Javascript empty migration file when option is passed`) and it doesn't seem to be related to the fix I made. I would appreciate some guidance.

That sounds possibly related to #11653 which has been fixed now. Tests are passing in this PR.

LeviHeber commented:
I made the following fixes:
1. Moved the test to its proper place in the eager-relations feature tests.
2. Updated the fix to work with `OrmUtils.mergeDeep` instead of `ObjectUtils.assign`

naorpeled commented:
@coderabbitai review

alumni commented:
I think the issue that we have could possibly be a bit earlier, when we build the query. It seems we select the user twice - maybe once because it's an eagerly loaded relation and another time because it was requested by the user.

In the test we have:

```ts
await connection.manager.findOne(
    Editor,
    {
        where: { userId: 1 },
        relations: {
            user: {
                nestedProfile: true,
            },
        },
    },
)
```

which generates the query:

```sql
SELECT
    "Editor"."userId" AS "Editor_userId",
    "Editor"."postId" AS "Editor_postId",
    "Editor__Editor_user"."id" AS "Editor__Editor_user_id",
    "Editor__Editor_user"."firstName" AS "Editor__Editor_user_firstName",
    "Editor__Editor_user"."lastName" AS "Editor__Editor_user_lastName",
    "Editor__Editor_user"."deletedAt" AS "Editor__Editor_user_deletedAt",
    "Editor__Editor_user"."profileId" AS "Editor__Editor_user_profileId",
    "Editor__Editor_user"."nestedProfileId" AS "Editor__Editor_user_nestedProfileId",
    "Editor__Editor_user__Editor__Editor_user_nestedProfile"."id" AS "Editor__Editor_user__Editor__Editor_user_nestedProfile_id",
    "Editor__Editor_user__Editor__Editor_user_nestedProfile"."about" AS "Editor__Editor_user__Editor__Editor_user_nestedProfile_about",
    "Editor__Editor_user__Editor__Editor_user_profile"."id" AS "Editor__Editor_user__Editor__Editor_user_profile_id",
    "Editor__Editor_user__Editor__Editor_user_profile"."about" AS "Editor__Editor_user__Editor__Editor_user_profile_about",
    "Editor__user"."id" AS "Editor__user_id",
    "Editor__user"."firstName" AS "Editor__user_firstName",
    "Editor__user"."lastName" AS "Editor__user_lastName",
    "Editor__user"."deletedAt" AS "Editor__user_deletedAt",
    "Editor__user"."profileId" AS "Editor__user_profileId",
    "Editor__user"."nestedProfileId" AS "Editor__user_nestedProfileId",
    "Editor__user__profile"."id" AS "Editor__user__profile_id",
    "Editor__user__profile"."about" AS "Editor__user__profile_about"
FROM
    "editor" "Editor"
    LEFT JOIN "user" "Editor__Editor_user" ON "Editor__Editor_user"."id" = "Editor"."userId"
    AND ("Editor__Editor_user"."deletedAt" IS NULL)
    LEFT JOIN "profile" "Editor__Editor_user__Editor__Editor_user_nestedProfile" ON "Editor__Editor_user__Editor__Editor_user_nestedProfile"."id" = "Editor__Editor_user"."nestedProfileId"
    LEFT JOIN "profile" "Editor__Editor_user__Editor__Editor_user_profile" ON "Editor__Editor_user__Editor__Editor_user_profile"."id" = "Editor__Editor_user"."profileId"
    LEFT JOIN "user" "Editor__user" ON "Editor__user"."id" = "Editor"."userId"
    AND ("Editor__user"."deletedAt" IS NULL)
    LEFT JOIN "profile" "Editor__user__profile" ON "Editor__user__profile"."id" = "Editor__user"."profileId"
WHERE
    ((("Editor"."userId" = 1)))
    AND (
        "Editor"."userId" = 1
        AND "Editor"."postId" = 1
    )
```

We have both `Editor__Editor_user`+`Editor__Editor_user__Editor__Editor_user_profile` and `Editor__user`+`Editor__user__profile`.

I'll check tomorrow to see if this is intended.
